eventsource-client/src/client.rs

```diff
@@ -517,30 +517,49 @@ where
                             }
                         }
 
-                        self.as_mut().reset_redirects();
-                        self.as_mut().project().state.set(State::New);
-
-                        return Poll::Ready(Some(Err(Error::UnexpectedResponse(
+                        let error = Error::UnexpectedResponse(
                             Response::new(resp.status(), resp.headers().clone()),
                             ErrorBody::new(resp.into_body()),
-                        ))));
+                        );
+
+                        if !*retry {
+                            self.as_mut().project().state.set(State::StreamClosed);
```

Comment:
Instead of introducing a new state as well, we are going to continue using `StreamClosed` as the only state which represents that the stream is exhausted and cannot be further consumed.